### ISSUE TITLE
Serializable for webworkers

### ISSUE THREAD

**Issue by natew** [facebook/immutable-js#336]
Wondering about the possibility of this. For now I am just doing data processing on the webworker, but seems like I need to pass back the raw JSON because immutable objects aren't stringify-able. Wondering, whats the feasibility of pushing all of the immutable stuff into a webworker given it needs to be able to be passed back as a string?

**Comment by leebyron** [comment]
Great suggestion. I need to look into web worker serialization a bit more to figure out the best path forward. Today there is not a clear way to serialize and deserialize Immutable collections.

**Comment by pedroteixeira** [comment]
I think one could use transit-js as mechanism to serialize/deserialize:
https://github.com/cognitect/transit-js
https://gist.github.com/Tvaroh/52efbe8f4541ca537908

**Comment by natew** [comment]
I thought the main problem would be that immutable collections have functions attached to them, which don't seem to be serializable. But typing that out makes me think that can't be right, you should be able to have the functions held separately. In the end it's all data.

@pedroteixeira that gist looks really interesting. Have you gotten it working? I already have webpack compiling requests inside webworkers working nicely, and immutable.js with flux (but not split into workers). If I could get the final piece (webworkers + immutable) I'd love to release a complete package that encompassed flux + immutable + webworkers + requests as a sort of super-data package for React.

**Comment by mrjjwright** [comment]
I was wondering today whether or not because of Relay/GraphQL there could be a increased need to serialize immutable data structures between client and server as well as between webworkers.   What would be interesting is to sync content between client and server, similar to how git does it.  Would there be a way to compares hashes between a local immutable structure and remote one and if they are different exchange content until they are the same.  With this approach the render immutable equality check would essentially become *has any content changed in the graph, locally or on the server".  I realize that Immutable.js is simply about immutable in memory persistent data structure in JavaScript but it's hard not to think about how syncing can radically change when using immutable data.

**Comment by leebyron** [comment]
Transit is a great tool, I definitely recommend using it. I'm also considering adding a very simple serialization API to the library

**Comment by xixixao** [comment]
+1 for serialization/deserialization API, I guess I'll go with type+JSON dump of toJS() for now.

**Comment by nmn** [comment]
+1 from me as well.
transit-js still uses JSON for the js implementation, so I'm not sure how well it would work.

Also, one big problem is that as soon as you serialise immutable data, you immediately lose all the benefits of structural sharing. I don't know nearly enough to be able to actually implement this, but the ideal way to pass data between the main thread and a web worker should work something like this:

1. Use SharedArrayBuffers to store all the data of an immutable data structure. (easier said than done. These array buffers are typed, and it will be really hard to store variable sized string etc.)
2. the immutable 'object' is just an object which stores the basic structures and indexes to the SharedArrayBuffer to know where to read what.
3. When passing data to a web worker, you can give it access to the sharedArrayBuffer and serialise and pass the immutable data structure that points to it.

Now, as long as no one directly tries to mutate the values on the Array Buffer, and uses the library, data will only be appended to Array Buffer and never mutated.

This would be essentially the same semantics as multi-threading in clojure. The one big difference here being that we pretty much need to manage memory manually. (because any shared memory needs to be in SharedArrayBuffers)

So I'm not sure if this is possible/feasible at all. But I wanted to write it down, just in case someone smart like @leebyron thinks its easy!

---
Some other ideas:
build an extension on top of JSON that has some sort of structural sharing built-in. (Falcor seems to be doing something similar) gzip takes care of duplicate data to a big extent, but it would be nice, never to have to generate a huge string in the first place

- TC39 should be pressured into accepting the proposal for immutable data in Javascript. Then we can put pressure to make passing these data structures to web-workers fast using memory sharing, since immutable data is read only, and hence, safe for multi-threading.

**Comment by glenjamin** [comment]
I've just published an early version of a module which provides the transit integration https://github.com/glenjamin/transit-immutable-js

I believe transit does make it possible to support internal structural sharing when serialising, but I haven't attempted to look into this yet.

**Comment by facebook-github-bot-6** [comment]
Thank you for reporting this issue and appreciate your patience. We've  notified the core team for an update on this issue. We're looking for a response within the next 30 days or the issue may be closed.

**Comment by piaolingxue** [comment]
+1

**Comment by rgbkrk** [comment]
> Today there is not a clear way to serialize and deserialize Immutable collections.

I'm actually looking to be able to write Immutable JS structures to disk, from an Electron app (in particular, the Jupyter/IPython notebook). This means I have a combined need of both streaming and serializing.

> I'm also considering adding a very simple serialization API to the library

@leebyron Is this still in the works to target for what I'm working on (release for us won't be a while).

**Comment by nmn** [comment]
@rgbkrk Technically speaking you CAN serialize immutable data. You just can't serialize the structural sharing that you get with immutable. That is an extremely hard problem to solve and I wouldn't expect any resolution on that in the near future.

**Comment by leebyron** [comment]
If you're looking for a serialization method you can use today, I would strongly suggest using transit.js. It's a great library for serialization and you can find write ups on using it with immutable.js.

Better methods for serialization are on the roadmap, but it's unlikely that we will serialize the structural sharing itself, as that can have very tricky implications. Instead we will do something fairly simple and related to what transit.js does

**Comment by glenjamin** [comment]
As mentioned above, https://github.com/glenjamin/transit-immutable-js will take care of the glue code between this and transit.

**Comment by rgbkrk** [comment]
Transit looks like the right case for webworkers, though the format is not the raw JSON I need to write out. I actually want to keep the `toJS()` structure and write to disk with json.

Currently it looks like this:

```javascript
fs.writeFile('plots.ipynb', JSON.stringify(notebook.toJS()), 'utf8', errCallback);
```

`notebook` above will end up being 10-25 MB at times

**Comment by rgbkrk** [comment]
Feel a little silly now that I've noticed that immutable objects can be passed straight to `JSON.stringify`. Thanks for bearing with me. It would be nice for me to be able to stream these directly to a file while being stringified; this will do though.

**Comment by runn1ng** [comment]
Hm. I was hit by this recently :(

It's sad that immutable objects are hard to put to webworkers, when immutable objects (in principle) should be *easier* to share - no need to copy, deal with write access, etc.

What's funny is that I cannot even mark it as Transferable - in principle, immutable object should be Transferable by default :)

Yeah, I can serialize and deserialize, but that triples the memory. I have a large immutable.js object, I cannot afford to serialize and deserialize again, it takes a long time and too much memory. Again it's ironic, since immutability should *help* to run things in parallel. But in this case, it makes things go more complex :(

Maybe if immutable became first-class citizen in browser :D haha, probably not.

**Comment by runn1ng** [comment]
....and it seems @nmn had exactly the same ideas, in 2015. :D ok

**Comment by leebyron** [comment]
Unfortunately this is still an open challenge, but I'm closing this aging issue

**Comment by runn1ng** [comment]
wontfix? :( ok then

**Comment by nmn** [comment]
I don't think it is fixable with today's technology.
